Introducing transverse updates through shared context

n1k0 opened the pull request:
The pages init and update functions must now return a tuple of `(Model, Session, Cmd Msg)`, allowing to update transverse app data right from within a page.

A new `Store` data structure is attached to a `Session` and aimed at storing the data to store across browser restarts (typically in localStorage). 

Requesting feedback from @magopian too :)

n1k0 committed:
Add Context.

n1k0 committed:
Page to return a Context.

fbentz reviewed (approved):
LGTM ! :+1:

n1k0 committed:
Rename Context to Shared /cc @fbentz

Natim reviewed (commented):
I like the fact that you explicitly pass the shared record to the update functions. However I am wondering why you decided to create yet another level that encapsulate session. Why create Shared rather than use Session directly?

n1k0 commented:
That's a good remark, I like to think of the Session as something serializable to some datastore, eg. localStorage. Maybe we could rename *Shared* to *Session* and *Session* to *SessionStore*?

magopian commented:
LGTM also, and I have the same question as @Natim : I don't see what having an extra layer gives us?

magopian commented:
> Session to SessionStore?

I guess your point is "let's nudge the devs to have a dedicated, separated state for something that will end up serialized. It makes sense I guess, and doesn't do any harm having here.
I'm not sure it's needed in a "minimal app" though, so not sure it makes sense to have it in here. I don't have a strong opinion about that.

I think my point is: we could very easily keep the PR as it is (`shared` is well named), and remove everything related to the (currently empty) `Session`, to make it as minimal and straightforward as possible, with the smallest amount of cruft.

n1k0 committed:
Session serialized data store implementation + ports.

/cc @Natim @magopian

n1k0 commented:
I've kept the Session record and added a Store one. I've also added Ports, encoders and decoders to expose a complete solution for persisting information across browser restarts. I feel this is useful, thoughts?

Natim commented on skeleton/public/index.html:
```diff
@@ -9,8 +9,26 @@
+    // Ensure session is refreshed when it changes in another tab/window
+    window.addEventListener("storage", (event) => {
+      if (event.storageArea === localStorage && event.key === storeKey) {
+        app.ports.storeChanged.send(event.newValue);
```
Why not use JSON.parse here since you make the stringify on this side as well? You can then use `decodeValue` on a `Decode.Value` to decode on the Elm side rather than `decodeString`

n1k0 replied:
Grrr, my comment was mangled. I was saying, I don't want to rely on JSON.parse and JSON.stringify js side because that would mean handling a try/catch for the error case. I've updated the code to always send raw json strings to Elm and proceed with parsing it there.

n1k0 commented on skeleton/elm.json:
```diff
@@ -30,4 +30,4 @@
         }
     }
-}
+}
```
This is the standard formatting after `elm install XXX`, not sure I should override that

Natim reviewed (approved):
I like it much more like that indeed.

n1k0 committed:
Serialize session data entirely.

n1k0 commented:
I'm still not 100% happy with the patch:

- I really liked the `Shared` name, which is now gone. `Session` doesn't convey enough meaning that the values it carries are transverse to all pages and attached to global app state.
- I'm not too fond of the API for updating the session within a page update:

```elm
update : Session -> Msg -> Model -> ( Model, Session, Cmd Msg )
update ({ store } as session) msg model =
    let
        newCount =
            model + 1

        newStore =
            { store | counter = newCount }
    in
    case msg of
        Inc ->
            ( newCount
            , { session | store = newStore }
            , newStore |> Session.serializeStore |> Ports.saveStore
            )
``` 

I have a feeling we could come with a better API where we could both update a store field value and trigger a save operation at once. I still need to think abou it though, maybe it's not that much of a brilliant idea.

n1k0 commented:
I have this patch where I compare the session store value retrieved from a page update with the previous session store value, and trigger a save command when it has changed:

```diff
diff --git a/skeleton/src/Main.elm b/skeleton/src/Main.elm
index 737c844..536cc3f 100644
--- a/skeleton/src/Main.elm
+++ b/skeleton/src/Main.elm
@@ -88,9 +88,16 @@ update msg ({ page, session } as model) =
             let
                 ( newModel, newSession, newCmd ) =
                     subUpdate session subMsg subModel
+
+                storeCmd =
+                    if session.store /= newSession.store then
+                        newSession.store |> Session.serializeStore |> Ports.saveStore
+
+                    else
+                        Cmd.none
             in
             ( { model | session = newSession, page = toModel newModel }
-            , Cmd.map toMsg newCmd
+            , Cmd.map toMsg (Cmd.batch [ newCmd, storeCmd ])
             )
     in
     case ( msg, page ) of
diff --git a/skeleton/src/Page/Counter.elm b/skeleton/src/Page/Counter.elm
index f267859..e34a06d 100644
--- a/skeleton/src/Page/Counter.elm
+++ b/skeleton/src/Page/Counter.elm
@@ -35,7 +35,7 @@ update ({ store } as session) msg model =
         Inc ->
             ( newCount
             , { session | store = newStore }
-            , newStore |> Session.serializeStore |> Ports.saveStore
+            , Cmd.none
             )
``` 

I'm not sure it's a good idea, though it works. It feels a little too magic for my taste. Thoughts? /cc @magopian @Natim

n1k0 committed:
Automatically persist updated session store data.

n1k0 commented:
I've pushed that strategy in https://github.com/allo-media/elm-kitchen/pull/12/commits/72393ab01d1ab1fe5f643551c70637fb51627e5f

n1k0 commented:
Kinda like the feel actually

![peek 2018-11-29 16-27](https://user-images.githubusercontent.com/41547/49232197-c7d49380-f3f3-11e8-92b8-eb0c510322d5.gif)

n1k0 commented:
No feedback is feedback, I'm landing this.
